ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt

```diff
@@ -281,6 +302,53 @@ class RuleEvalTest {
+        assertEquals(1f, eval("""gte(ip("192.168.0.255"), ip("192.168.0.255"))"""))
+        assertEquals(1f, eval("""lt(ip("255.0.255.255"), ip("255.1.0.0"))"""))
+        assertEquals(1f, eval("""lt(ip("9.255.255.255"), ip("10.0.0.0"))"""))
+
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.0.0")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.-1")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.256")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.a")""") }
+    }
+
+    @Test
+    fun cidr() {
+        assertEquals(0f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.224.0"))"""))
+        assertEquals(0f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.221.255"))"""))
+        assertEquals(0f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.200.10"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.16"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.232.255"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.255.16"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.11/32"), ip("192.167.233.12"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.11"), ip("192.167.233.12"))"""))
+
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.223.255"))"""))
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.222.0"))"""))
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.222.188"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.15"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.0"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.6"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.11/32"), ip("192.167.233.11"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.11"), ip("192.167.233.11"))"""))
+        assertEquals(1f, eval("""contains(cidr("0.0.0.0/0"), ip("0.0.0.0"))"""))
+        assertEquals(1f, eval("""contains(cidr("0.0.0.0/0"), ip("255.255.255.255"))"""))
+        assertEquals(1f, eval("""contains(cidr("0.0.0.0/0"), ip("192.168.44.41"))"""))
+        assertEquals(1f, eval("""contains(cidr("0.0.0.0/0"), ip("41.173.112.199"))"""))
+
+        assertEquals(0f, eval("""contains([cidr("254.200.222.210/23")], ip("254.200.222.25"))"""))
+        assertFailsWith<IllegalArgumentException> {
+            eval("""contains(cidr("254.200.222.210/23"))""")
+        }
+        assertFailsWith<IllegalArgumentException> {
+            eval("""contains([cidr("254.200.222.210/23", "254.200.224.0")])""")
+        }
+    }
```

Comment:
I've just realized that both `cird()` and `ip()` tests have somewhat of a large scope. Not only do they test the respective functions, they also test `contains` and `gt` (and friends). It would be great to make them more focused by testing specifically the outputs of each, and delegating a few of the ones that compose with other function to `composition`.

Reply:
I'm thinking of changing `private sealed class RuleExpr<out T>` to public, this will help us test the IP classes without having to do composition

Currently, the `ip` and `cidr` has no exposed public method, so either 
1. We made it public, or
2. Create a test helper method within RuleEval.kt

Alternative 1 is cleaner in my opinion

Reply:
What is the blocker to testing via eval? Precision?

Reply:
Precision is one of the considerations, but the bigger blocker to testing is the `cidr` returns range, which can't be converted with `eval`. This is the reason I'm using `contains` to run the test

This challenge is even more prominent on IPv6 implementation proposal

```kotlin
return when (val result = RuleGrammar.parseToEnd(formula).eval(env)) {
    is Boolean -> if (result) 1f else 0f
    is Number -> result.toFloat()
    is String -> result.runCatching { toFloat() }.recoverCatching { 0f }.getOrThrow()
    else -> 0f
}
```

This got me thinking, what about extracting all the IP related functions to a data class, and testing that class separately outside RuleEval

It might be a bit excessive, but it will decouple RuleEval logic from IP related logic

Reply:
Gotcha. Let's keep the logic within that file as it's all connected and only used there, but the approach seems sound! That would benefit most other tests, too. One suggestion is to use `internal` instead of `public` since these modules are not meant to be used externally.

Reply:
Sounds good! 😊👍